src/conv-core/memory-record.C

```diff
@@ -0,0 +1,329 @@
+// Heap Record-Replay
+// By Evan Ramos
+// Generates an MPI program that repeats the exact sequence of heap operations observed on each PE.
+
+#include <cstdint>
+#include <cstdio>
+#include <unordered_map>
+
+// raw string literals containing the output program
+
+static const char prologue[] =
+R"code(
+#include <stdlib.h>
+#include <stdio.h>
+#include "mpi.h"
+
+int main(int argc, char ** argv)
+{
+  MPI_Init(&argc, &argv);
+
+  int p;
+  MPI_Comm_size(MPI_COMM_WORLD, &p);
+
+  if (%d != p)
+    puts("This recording was intended for %d ranks.");
+
+  int rank;
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+
+  switch (rank)
+  {
+)code";
+
+static const char body[] =
+R"code(
+    case %d:
+    {
+#include "heap-replay-%d.cpp"
+      break;
+    }
+)code";
+
+static const char epilogue[] =
+R"code(
+    default:
+      break;
+  }
+
+  MPI_Finalize();
+  return 0;
+}
+)code";
+
+// end raw string literals
+
+static char initialized;
+struct MemoryRecord {
+  std::unordered_map<uintptr_t, unsigned long long> ptrs;
+  FILE * output;
+  unsigned long long count;
+  char guard;
+};
+CpvStaticDeclare(MemoryRecord, state);
+
+static void meta_init(char **argv)
+{
+  CpvInitialize(MemoryRecord, state);
+
+  const int mype = CmiMyPe();
+
+  char outputname[64];
+  sprintf(outputname, "heap-replay-%d.cpp", mype);
+  FILE * output = fopen(outputname, "wb");
+  if (output == nullptr)
+    CmiAbort("Could not open heap-replay-%d.cpp!", mype);
+  CpvAccess(state).output = output;
+
+  if (CmiMyRank() == 0)
+    initialized = 1;
+
+  CmiNodeAllBarrier();
+
+  if (mype == 0)
+  {
+    const int numpes = CmiNumPes();
+
+    FILE * mainsrc = fopen("heap-replay.cpp", "wb");
```

Comment:
should probably be 'w' instead of 'wb' (same for the other fopen())

Reply:
Done